File: ClassEval_T/cpp/solution/AutomaticGuitarSimulator.cpp

#include <sstream>
#include <iostream>

#include <string>
#include <vector>

struct PlayItem {
    std::string Chord;
    std::string Tune;
};

class AutomaticGuitarSimulator {
public:
    AutomaticGuitarSimulator(const std::string& text);
    std::vector<PlayItem> interpret(bool display = false);
    std::string format_display(const std::string& key, const std::string& value);

private:
    std::string play_text;
};

AutomaticGuitarSimulator::AutomaticGuitarSimulator(const std::string& text) : play_text(text) {}

std::vector<PlayItem> AutomaticGuitarSimulator::interpret(bool display) {
    std::vector<PlayItem> play_list;
    if (play_text.empty()) {
        return play_list;
    }

    std::istringstream iss(play_text);
    std::string play_seg;

    while (iss >> play_seg) {
        size_t pos = 0;
        while (pos < play_seg.size() && isalpha(play_seg[pos])) {
            pos++;
        }

        std::string play_chord = play_seg.substr(0, pos);
        std::string play_value = play_seg.substr(pos);

        PlayItem item = { play_chord, play_value };
        play_list.push_back(item);

        if (display) {
            std::cout << format_display(play_chord, play_value) << std::endl;
        }
    }
    return play_list;
}

std::string AutomaticGuitarSimulator::format_display(const std::string& key, const std::string& value) {
    return "Normal Guitar Playing -- Chord: " + key + ", Play Tune: " + value;
}

File: ClassEval_T/java/solutuon/AutomaticGuitarSimulator.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public class AutomaticGuitarSimulator {
    private String playText;

    public AutomaticGuitarSimulator(String text) {
        this.playText = text;
    }

    public List<ChordTune> interpret(boolean display) {
        if (playText == null || playText.trim().isEmpty()) {
            return null;
        }

        List<ChordTune> playList = new ArrayList<>();
        String[] playSegs = playText.split(" ");
        for (String playSeg : playSegs) {
            if (playSeg.trim().isEmpty()) {
                continue;
            }
            int pos = 0;
            for (char ele : playSeg.toCharArray()) {
                if (Character.isLetter(ele)) {
                    pos++;
                    continue;
                }
                break;
            }
            String playChord = playSeg.substring(0, pos);
            String playValue = playSeg.substring(pos);
            playList.add(new ChordTune(playChord, playValue));
            if (display) {
                display(playChord, playValue);
            }
        }
        return playList;
    }

    public String display(String key, String value) {
        return String.format("Normal Guitar Playing -- Chord: %s, Play Tune: %s", key, value);
    }

    public static class ChordTune {
        private String chord;
        private String tune;

        public ChordTune(String chord, String tune) {
            this.chord = chord;
            this.tune = tune;
        }

        public String getChord() {
            return chord;
        }

        public String getTune() {
            return tune;
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            ChordTune that = (ChordTune) obj;
            return chord.equals(that.chord) && tune.equals(that.tune);
        }

        @Override
        public int hashCode() {
            int result = chord.hashCode();
            result = 31 * result + tune.hashCode();
            return result;
        }
    }
}